contracts/base/BaseLiquidityManagement.sol

```diff
@@ -96,39 +110,76 @@ contract BaseLiquidityManagement is IBaseLiquidityManagement, SafeCallback {
-            callerDelta = toBalanceDelta(0, callerDelta.amount1());
-        } else {
-            position.tokensOwed0 = 0;
-        }
+        // Calculate the portion of the liquidityDelta that is attributable to the caller.
+        // We must account for fees that might be owed to other users on the same range.
+        (uint256 feeGrowthInside0X128, uint256 feeGrowthInside1X128) =
+            manager.getFeeGrowthInside(range.poolKey.toId(), range.tickLower, range.tickUpper);
 
-        if (callerDelta.amount1() > 0) {
-            position.tokensOwed1 += uint128(callerDelta.amount1());
-            range.key.currency1.take(manager, address(this), uint128(callerDelta.amount1()), true);
-            callerDelta = toBalanceDelta(callerDelta.amount0(), 0);
-        } else {
-            position.tokensOwed1 = 0;
+        BalanceDelta callerFeesAccrued = FeeMath.getFeesOwed(
+            feeGrowthInside0X128,
+            feeGrowthInside1X128,
+            position.feeGrowthInside0LastX128,
+            position.feeGrowthInside1LastX128,
+            position.liquidity
+        );
+
+        // Calculate the accurate tokens owed to the caller.
+        // If the totalFeesAccrued equals the callerFeesAccrued then the total owed to the caller is just the liquidityDelta.
+        // If the totalFeesAccrued is greater than the callerFeesAccrued, we must account for the difference.
+        (int128 callerDelta0, int128 callerDelta1) = totalFeesAccrued != callerFeesAccrued
+            ? _calculateCallerDeltas(liquidityDelta, totalFeesAccrued, callerFeesAccrued)
+            : (liquidityDelta.amount0(), liquidityDelta.amount1());
+
+        // Update position storage, sanitizing the tokensOwed values based on the actual funds owed to the caller.
+        // if callerDelta > 0, then even after re-investing old fees, the caller still has some amount to collect that were not added into the position so they are accounted.
+        // if callerDelta <= 0, then tokensOwed0 and tokensOwed1 should be zero'd out as all fees were re-invested into a new position.
+        uint128 tokensOwed0 = 0;
+        uint128 tokensOwed1 = 0;
+        if (callerDelta0 > 0) {
+            tokensOwed0 = uint128(callerDelta0);
+            callerDelta0 = 0;
+        }
+        if (callerDelta1 > 0) {
+            tokensOwed1 = uint128(callerDelta1);
+            callerDelta1 = 0;
         }
 
-        return callerDelta;
+        position.updateTokensOwed(tokensOwed0, tokensOwed1);
+        position.addLiquidity(liquidityToAdd);
+        position.updateFeeGrowthInside(feeGrowthInside0X128, feeGrowthInside1X128);
+
+        return toBalanceDelta(callerDelta0, callerDelta1);
+    }
+
+    // Returns the new sanitized delta for the caller.
+    function _calculateCallerDeltas(
+        BalanceDelta liquidityDelta,
+        BalanceDelta totalFeesAccrued,
+        BalanceDelta callerFeesAccrued
+    ) private pure returns (int128 callerDelta0, int128 callerDelta1) {
```

Comment:
I think we're missing `position.tokensOwed` in the callerDelta calculation?

Reply:
why do we need that?